enhancements/cluster-logging/elasticsearch-v2-api-and-infrastructure.md

```diff
@@ -0,0 +1,310 @@
+
+* Determine a v2 elasticsearch CRD API
+
+* Discuss upgrade strategies from v1 to v2
+
+We will be successful when:
+
+* Users of Elasticsearch Operator no longer need to specify their node roles and count
+
+### Non-Goals
+
+* N/A
+
+### Current workflow vs Proposed
+
+Currently, a consumer like Jaeger or Cluster Logging is responsible for the following for their ES cluster:
+
+1. Creating and updating their elasticsearch CR(s) so that EO can manage their cluster
+    * This includes breaking up nodes by roles
+    * Specifying (usually pass through) of node requirements
+
+With this proposal the workflow would be:
+
+1. Create their elasticsearch CR specifying a particular sized cluster (based on t-shirt sizes)
+
+Current specification:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  managementState: "Managed"
+  nodeSpec:
+    resources:
+      limits:
+        memory: 1Gi
+      requests:
+        cpu: 100m
+        memory: 1Gi
+  nodes:
+  - nodeCount: 1
+    roles:
+    - client
+    - data
+    - master
+    storage:
+      storageClassName: gp2
+      size: 10Gi
+  redundancyPolicy: ZeroRedundancy
+  indexManagement:
+    policies:
+    - name: infra-policy
+      pollInterval: 1m
+      phases:
+        hot:
+          actions:
+            rollover:
+              maxAge:   2m
+        delete:
+          minAge: 5m
+    mappings:
+    - name:  infra
+      policyRef: infra-policy
+      aliases:
+      - infra
+      - logs.infra
+```
+
+Proposed:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  storage:
+    storageClassName: gp2
+    size: 10Gi
+  size: 0x.medium
+  indexManagement:
+    infraPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge:   2m
+      delete:
+        minAge: 5m
+    appPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge: 2m
+      delete:
+        minAge: 5m
+```
+
+## Proposal
+
+Currently the Elasticsearch CR is overly complicated and allows for too much granular control over the data nodes, which places the expectation
+of expertise on the user rather than containing it to the operator (which is where it should reside). This proposal seeks to outline a v2 to provide
+a more simple means of interracting with the Elasticsearch Operator to ask for a managed Elasticsearch cluster that can be better
+configured for performance.
+
+### User Stories
+
+#### As an user of EO, I want to have to provide minimal information for my cluster to be created for me
+
+As an user I want to just be able to provide information such as a t-shirt size (which is outlined further for me) and my storage information
+and have an elasticsearch cluster spun up for me so that I do not need to be an expert and configure my node roles and node resources.
```

Comment:
This is a great user story ... up until we give them t-shirt size, storage size, and retention.

I think we might want to consider ask the user for:

- max clients to support
- max # of separate indices
- max data retention times

And then use that to compute a t-shirt size that matches.  We can refine the exact inputs, but there is a formula of sorts from all the recommendations that are out there.

Reply:
the idea was actually that we would provide a "calculator" that would sit out there for 5 minutes to determine how much logs are generated and based on that extrapolate how many logs are created a day.

I think the most appropriate of those would be "max data retention times", however we still need to have an understanding of what size of each node will the customer allow ES to use -- hence the t-shirt size.

The shirt size doesn't need to strictly be cpu/memory but it is a templated way to let users tell us what size nodes to use and then let us scale up from there to keep the cluster in a sane state.